crates/ra_hir_def/src/nameres.rs

```diff
@@ -100,19 +102,77 @@ impl std::ops::Index<LocalModuleId> for CrateDefMap {
     }
 }
 
+#[derive(Debug, PartialEq, Eq, Clone, Copy, Hash)]
+pub enum ModuleOrigin {
+    /// It should not be `None` after collecting definitions.
+    Root(Option<FileId>),
+    /// Note that non-inline modules, by definition, live inside non-macro file.
+    File(AstId<ast::Module>, FileId),
+    Inline(AstId<ast::Module>),
+}
+
+impl Default for ModuleOrigin {
+    fn default() -> Self {
+        ModuleOrigin::Root(None)
+    }
+}
+
+impl ModuleOrigin {
+    pub fn root(file_id: FileId) -> Self {
+        ModuleOrigin::Root(Some(file_id))
+    }
+
+    pub fn not_sure_file(file: Option<FileId>, module: AstId<ast::Module>) -> Self {
+        match file {
+            None => ModuleOrigin::Inline(module),
+            Some(file) => ModuleOrigin::File(module, file),
+        }
+    }
+
+    pub fn not_sure_mod(file: FileId, module: Option<AstId<ast::Module>>) -> Self {
```

Comment:
Let's make all constructors private or `pub(crate)`, it doesn't feel right that they are world-visible

Reply:
And looks like some of those constructors are not used at all?

Reply:
> Let's make all constructors private or `pub(crate)`, it doesn't feel right that they are world-visible

They are used in `ra_hir`

Reply:
> And looks like some of those constructors are not used at all?

I have already deleted one, not sure about the others

Reply:
Hm, I see zero usages for `not_usre_mod`, and a single usage, in this crate, for `not_sure_file`

Reply:
> Hm, I see zero usages for `not_usre_mod`, and a single usage, in this crate, for `not_sure_file`

Well I thought it's `ModuleSource`. Now it's clear.